src/sage/rings/dirichlet_series_ring_element.py

```diff
@@ -0,0 +1,404 @@
+            1 + 2*2^-s + 3*3^-s + 4*4^-s + O(5^-s)
+            sage: g.parent()
+            Dirichlet Series Ring over Integer Ring with fixed precision 5
+        """
+        parent = self.parent()
+        if new_parent:
+            from sage.rings.dirichlet_series_ring import DirichletSeriesRing
+            parent = DirichletSeriesRing(parent.base_ring(), prec, parent.is_sparse())
+        return parent({i: j for i,j in self.coefficients().items() if i < prec})
+
+class DirichletSeries_dense(DirichletSeries_generic):
+    def __init__(self, parent, data=None):
+        base_ring = parent.base_ring()
+        zero = base_ring(0)
+        self.data = [zero for _ in range(parent.precision())]
+        DirichletSeries_generic.__init__(self, parent, data)
+
+    def coefficients(self):
+        return {i: j for i,j in enumerate(self.data) if j}
+
+    def _add_(self, other):
+        parent = self.parent()
+        return parent([self[i] + other[i] for i in range(1, len(self.data))])
+
+    def _sub_(self, other):
+        parent = self.parent()
+        return parent([self[i] - other[i] for i in range(1, len(self.data))])
+
+    def _mul_(self, other):
+        return self.dense_times_generic(other)
+
+    def dense_times_generic(self, other):
+        """
+        Multiply a dense Dirichlet series by a generic Dirichlet series.
+        """
+        parent = self.parent()
+        base_ring = self.base_ring()
+        ans = [base_ring.zero() for i in range(1, parent.precision())]
+        for i, x in other.coefficients().items():
+            for j in range(1, (parent.precision()-1)//i + 1):
+                y = self[j]
+                if y:
+                    ans[i*j-1] += x*y
+        return parent(ans)
+
+class DirichletSeries_sparse(DirichletSeries_generic):
+    def __init__(self, parent, data=None):
+        self.data = {}
+        DirichletSeries_generic.__init__(self, parent, data)
+
+    def coefficients(self):
+        return {i: j for i,j in self.data.items() if j}
+
+    def _add_(self, other):
+        parent = self.parent()
+        base_ring = self.base_ring()
+        out = dict(self.data)
+        for i,j in other.data.items():
+            out[i] = out[i] + j if i in out else j
+        return parent(out)
+
+    def _sub_(self, other):
+        parent = self.parent()
+        base_ring = self.base_ring()
+        out = {i: j for i,j in self.data.items()}
+        for i,j in other.data.items():
+            out[i] = out[i] - j if i in out else -j
+        return parent(out)
+
+    def _mul_(self, other):
+        parent = self.parent()
+        base_ring = self.base_ring()
+        out = {}
+        prec = parent.precision()
+        for i, x in self.data.items():
+            for j, y in other.data.items():
+                k = i*j
+                if k < prec:
+                    out[k] = out[k] + x*y if k in out else x*y
+        return parent(out)
+
+class DirichletSeriesIterator:
```

Comment:
I don't see the need for this class. Why not just use the default iterator of the defining list?